File: oss/src/main/java/io/tn/oss/minio/MinioClientWrapper.java
package io.tn.oss.minio;

import io.minio.MinioClient;
import io.minio.messages.Bucket;
import io.tn.oss.FileArgs;
import io.tn.oss.abstracts.InMap;
import io.tn.oss.abstracts.Oss;
import io.tn.oss.abstracts.OutMap;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * oss 的 minio 实现
 *
 * @author TrueNine
 * @since 2023-02-20
 */
public non-sealed class MinioClientWrapper extends MinioClientCompartment implements Oss {
  public MinioClientWrapper(MinioClient client) {
    super(client);
  }

  @Override
  @SuppressWarnings("unchecked")
  public <T> T as(Class<T> instanceType) {
    return (T) super.getClient();
  }

  @Override
  public void makeDirs(String dirName) {
    createBucket(dirName);
  }

  @Override
  public void publicDir(String dir) {
    publicBucket(dir);
  }

  @Override
  public InMap upload(InputStream stream, FileArgs fileInfo) {
    var ins = putObject(fileInfo, stream);
    return ins(ins, stream);
  }

  @Override
  public InMap upload(InputStream stream, FileArgs fileInfo, Runnable afterExec) {
    var ins = upload(stream, fileInfo);
    afterExec.run();
    return ins;
  }

  @Override
  public InMap upload(InputStream stream, FileArgs fileArgs, Consumer<FileArgs> afterExec) {
    var ins = upload(stream, fileArgs);
    afterExec.accept(FileArgs.useStreamMap(ins));
    return ins;
  }

  @Override
  public OutMap download(OutputStream stream, FileArgs fileInfo) throws IOException {
    var outs = getObject(fileInfo, stream);
    outs.transferTo(stream);
    return outs(outs, stream);
  }

  @Override
  public OutMap download(Runnable beforeExec, OutputStream stream, FileArgs fileInfo) throws IOException {
    var outs = getObject(fileInfo, stream);
    beforeExec.run();
    outs.transferTo(stream);
    return download(stream, fileInfo);
  }

  @Override
  public OutMap download(Consumer<FileArgs> beforeExec, OutputStream stream, FileArgs fileInfo) throws IOException {
    var outs = getObject(fileInfo, stream);
    var wrapper = outs(outs, stream);
    beforeExec.accept(FileArgs.useStreamMap(wrapper));
    outs.transferTo(stream);
    return outs(outs, stream);
  }

  @Override
  public List<String> listFiles(String dir) {
    return super.listFiles(dir);
  }

  @Override
  public List<String> listDir() {
    return getBuckets().stream().map(Bucket::name)
      .collect(Collectors.toList());
  }
}

File: oss/src/main/kotlin/net/yan100/compose/oss/minio/MinioClientWrapper.kt
package net.yan100.compose.oss.minio

import io.minio.MinioClient
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.function.Consumer
import net.yan100.compose.oss.FileArgs
import net.yan100.compose.oss.InMap
import net.yan100.compose.oss.Oss
import net.yan100.compose.oss.OutMap

/**
 * oss 的 minio 实现
 *
 * @author TrueNine
 * @since 2023-02-20
 */
class MinioClientWrapper(
  private val minioClient: MinioClient,
  private val exposeUrl: String = "http://localhost:9000",
) : Oss, MinioClientAdaptor(minioClient, exposeUrl) {

  override fun getNative(): Any {
    return super.client
  }

  override val exposedBaseUrl: String
    get() = this.exposeUrl

  override fun makeDirs(dirName: String) {
    createBucket(dirName)
  }

  override fun existsDir(dirName: String): Boolean {
    return bucketExists(dirName)
  }

  override fun removeFile(fileInfo: FileArgs): Boolean {
    return removeObject(fileInfo)
  }

  override fun publicDir(dir: String) {
    publicBucket(dir)
  }

  override fun upload(stream: InputStream, fileInfo: FileArgs): InMap {
    val ins = putObject(fileInfo, stream)
    return ins(ins!!, stream)
  }

  override fun upload(
    stream: InputStream,
    fileInfo: FileArgs,
    afterExec: Runnable,
  ): InMap {
    val ins = upload(stream, fileInfo)
    afterExec.run()
    return ins
  }

  override fun upload(
    stream: InputStream,
    fileArgs: FileArgs,
    afterExec: Consumer<FileArgs>,
  ): InMap {
    val ins = upload(stream, fileArgs)
    afterExec.accept(FileArgs.useStreamMap(ins))
    return ins
  }

  @Throws(IOException::class)
  override fun download(stream: OutputStream, fileInfo: FileArgs): OutMap {
    val outs = getObject(fileInfo, stream)
    outs?.transferTo(stream)
    return outs(outs!!, stream)
  }

  @Throws(IOException::class)
  override fun download(
    beforeExec: Runnable,
    stream: OutputStream,
    fileInfo: FileArgs,
  ): OutMap {
    val outs = getObject(fileInfo, stream)
    beforeExec.run()
    outs?.transferTo(stream)
    return download(stream, fileInfo)
  }

  @Throws(IOException::class)
  override fun download(
    beforeExec: Consumer<FileArgs>,
    stream: OutputStream,
    fileInfo: FileArgs,
  ): OutMap {
    val outs = getObject(fileInfo, stream)
    val wrapper = outs(outs!!, stream)
    beforeExec.accept(FileArgs.useStreamMap(wrapper))
    outs.transferTo(stream)
    return outs(outs, stream)
  }
}
